File: src/oo/State.java
package oo;

interface State {
    String writeName(StateContext context, String name);
}

class StateContext {
    private State state = new LowerCaseState();

    void setState(State newState) {
        state = newState;
    }

    public void writeName(String name) {
        state.writeName(this, name);
    }
}

class LowerCaseState implements State {
    @Override public String writeName(StateContext context, String name) {
        context.setState(new MultipleUpperCaseState());
        return name.toLowerCase();
    }
}

class MultipleUpperCaseState implements State {
    private int count = 0;

    @Override public String writeName(StateContext context, String name) {
        if (++count > 1) {
            context.setState(new LowerCaseState());
        }
        return name.toUpperCase();
    }
}

File: src/oo/State.kt
package oo

class Context {
    private var state: State = LowerCase()

    internal fun setState(state: State) { this.state = state }

    fun echo(string: String) = state.echo(string, this)
}

interface State {
    fun echo(string: String, context: Context): String
}

class LowerCase : State {
    override fun echo(string: String, context: Context): String {
        val result = string.toLowerCase()
        context.setState(UpperCase())
        return result
    }
}

class UpperCase : State {
    private var data = 0

    override fun echo(string: String, context: Context): String {
        val result = string.toUpperCase()
        if (data > 0)
            context.setState(LowerCase())
        else
            data++
        return result
    }
}
